pkg/database/workouts.go

```diff
@@ -82,7 +83,29 @@ func NewWorkout(u *User, workoutType WorkoutType, notes string, filename string,
 	return &w, nil
 }
 
-func autoDetectWorkoutType(data *MapData) WorkoutType {
+func autoDetectWorkoutType(data *MapData, gpxContent *gpx.GPX) WorkoutType {
+	// If the GPX file mentions a workout type, use it
+	if len(gpxContent.Tracks) > 0 {
```

Comment:
You should make a new function for this; the code will be cleaner.

```golang
func (m *MapData) workoutTypeFromType() *WorkoutType {
  if len(gpxContent.Tracks) == 0 {
    return nil
  }

  firstTrack := &gpxContent.Tracks[0]

  switch strings.ToLower(firstTrack.Type) {
  case [...]
  default:
    return nil
  }  
}
```

and then:

```golang
  w := data.workoutTypeFromType()
  if w != nil {
    return *w // Not sure about the *w or &w, I always rely on my editor/compiler for this :)
  }

  if 3.6*data.AverageSpeedNoPause() > 15.0 {
  [...]
```